### PR TITLE
Stdlib rewriting

### PR DESCRIPTION
**Pull request by pascutto**
*Don't merge yet*: once we agree on the stdlib, we need to fix the tests.

### CODE REVIEW

File: src/stdlib/gospelstdlib.mli [hunk 1]
```diff
@@ -31,359 +31,576 @@
...
+
+(*@ function ({}) : 'a set *)
+(** [\{\}] is the empty set. *)
+
 module Set : sig
+  (*@ type 'a t = 'a set *)
+  (** An alias for ['a set]. *)
+
+  (*@ function compare (s s': 'a t) : int *)
+  (** A comparison function over sets. *)
```
Comment (n-osborne) [inline src/stdlib/gospelstdlib.mli:480]: Maybe add some precision about the semantic?

File: src/stdlib/gospelstdlib.mli [hunk 2]
```diff
@@ -31,359 +31,576 @@
...
+  (** [filter_map f s] is the set of all [v] such that [f x = Some v] for some
+      element [x] of [s]. *)
 
-  (*@ function sum (f:'a -> integer) (x: 'a set) : integer *)
+  (*@ function partition (f: 'a -> bool) (s: 'a t) : ('a t * 'a t) *)
+  (** [partition f s] is the pair of sets [(s1, s2)], where [s1] is the set of
+      all the elements of [s] that satisfy the predicate [f], and [s2] is the
+      set of all the elements of [s] that do not satisfy [f]. *)
 
+  (*@ function cardinal (s: 'a t) : int *)
```
Comment (n-osborne) [inline src/stdlib/gospelstdlib.mli:551]: Already defined line 518.

File: src/stdlib/gospelstdlib.mli [hunk 3]
```diff
@@ -31,359 +31,576 @@
...
+  (** [cardinal s] is the number of elements in [s]. *)
+
+  (*@ function elements (s: 'a t) : 'a list *)
+  (** [elements s] is the list of all elements in [s]. *)
+
+  (*@ function choose (s: 'a t) : 'a *)
```
Comment (n-osborne) [inline src/stdlib/gospelstdlib.mli:557]: Already defined line 521.

File: src/stdlib/gospelstdlib.mli [hunk 4]
```diff
@@ -31,359 +31,576 @@
...
-      ensures  forall i. 0 <= i < n -> result[i] = f i *)
+  (*@ function tl (s: 'a t) : 'a t *)
+  (** When [s] contains one or more elements, [tl s] is the sequence of the
+      elements of [s], starting at position 2. *)
+
+  (*@ function append (s s': 'a t) : 'a t *)
+  (** [append s s'] is [s ++ s']. *)
+
+  (*@ predicate mem (s: 'a seq) (x: 'a) *)
+  (** [mem s x] is true iff [x] is in [s]. *)
```
Comment (n-osborne) [inline src/stdlib/gospelstdlib.mli:179]: [mem s x] holds iff [x] is in [s].

File: src/stdlib/gospelstdlib.mli [hunk 5]
```diff
@@ -31,359 +31,576 @@
...
+      length of [t]. *)
 
-(** Lists
+  (*@ function map2 (f: 'a -> 'b -> 'c) (l: 'a t) (l': 'b t) : 'c t *)
+  (** [map2 f l l'] applies function [f] to all the elements of [l] and [l'],
+      and builds a list with the results returned by [f]. *)
 
-     Type 'a list, [] and (::) constructors are built-in *)
+  (*@ predicate for_all (f: 'a -> bool) (l: 'a t) *)
+  (** [for_all f l] checks if all elements of [l] satisfy the predicate [f]. *)
```
Comment (n-osborne) [inline src/stdlib/gospelstdlib.mli:264]: The documentation of some predicates/quantifier is written with `checks` and some other with `holds iff`. Maybe we should choose one.

**Review by n-osborne — COMMENTED** [review]
I think it would be good to have more consistency in the documentation (style).

As some modules have relatively the same interface, would is be better to have the functions always in the same order?

File: src/stdlib/gospelstdlib.mli [hunk 6]
```diff
@@ -31,359 +31,576 @@
...
+
+(*@ function succ (x: integer) : integer *)
+(*@ function pred (x: integer) : integer *)
+
+(*@ function (-_) (x: integer) : integer *)
+(*@ function (+) (x y: integer) : integer *)
+(*@ function (-) (x y: integer) : integer *)
+(*@ function ( * ) (x y: integer) : integer *)
+(*@ function (/) (x y: integer) : integer *)
+(*@ function mod (x y: integer) : integer *)
```
[the remark this answers is not in the document]
Comment (pascutto) [reply src/stdlib/gospelstdlib.mli:63]: Indeed, the Gospel parser does not let you chose arbitrary infix symbols, and `mod` is not (yet?) built-in.

File: src/stdlib/gospelstdlib.mli (hunk 5 above)
Comment (pascutto) [reply src/stdlib/gospelstdlib.mli:264]: `holds iff` is probably better, those are predicates, not functions, so they don't really check anything

**Comment by pascutto** [comment]
Thanks! All fixed now (besides the infix `mod`, I'll leave it for another PR since it's a bit beyond the scope of the stdlib)

**Pull request merged**